From 0617cd243c03dabab867f6417e52247d4aa56a11 Mon Sep 17 00:00:00 2001
From: "Ye.Li" <Ye.Li@freescale.com>
Date: Fri, 10 Jul 2015 19:13:53 +0800
Subject: [PATCH] MLK-11236 mtd: gpmi-nand: Fix nand runtime PM issue

Because of the delay of auto suspend, the nand clocks are delayed to
disable when calling the clk_set_rate. This causes the clk_set_rate
failed on some platforms like 6q/6qp, and finally lead the NAND not
working.

Signed-off-by: Ye.Li <Ye.Li@freescale.com>
---
 drivers/mtd/nand/gpmi-nand/gpmi-lib.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/mtd/nand/gpmi-nand/gpmi-lib.c b/drivers/mtd/nand/gpmi-nand/gpmi-lib.c
index 1fa46a2..b8ce976 100644
--- a/drivers/mtd/nand/gpmi-nand/gpmi-lib.c
+++ b/drivers/mtd/nand/gpmi-nand/gpmi-lib.c
@@ -959,9 +959,14 @@ static int enable_edo_mode(struct gpmi_nand_data *this, int mode)
 
 	nand->select_chip(mtd, -1);
 
+	pm_runtime_get_sync(this->dev);
+	clk_disable_unprepare(r->clock[0]);
 	/* [3] set the main IO clock, 100MHz for mode 5, 80MHz for mode 4. */
 	rate = (mode == 5) ? 100000000 : 80000000;
 	clk_set_rate(r->clock[0], rate);
+	clk_prepare_enable(r->clock[0]);
+	pm_runtime_mark_last_busy(this->dev);
+        pm_runtime_put_autosuspend(this->dev);	
 
 	/* Let the gpmi_begin() re-compute the timing again. */
 	this->flags &= ~GPMI_TIMING_INIT_OK;
-- 
1.8.0

